### PR TITLE
Developer Experience improvements

### LINKED ISSUE #98: Document developer workflow
ref: #92
/milestone baseline 
/assign

### LINKED ISSUE #107: Investigating using distroless/static for image building
/kind feature

ref: #92, https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/88/files#r259902819

**Describe the solution you'd like**
[A clear and concise description of what you want to happen.]

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

**Environment:**

- cluster-api-provider-azure version: v0.1.0-alpha.3
- Kubernetes version: (use `kubectl version`): 1.13.3
- OS (e.g. from `/etc/os-release`): Ubuntu 18.04

/help

### LINKED ISSUE #108: Clean up Makefile
/kind feature

ref: #92

**Describe the solution you'd like**
[A clear and concise description of what you want to happen.]

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

**Environment:**

- cluster-api-provider-azure version: v0.1.0-alpha.3
- Kubernetes version: (use `kubectl version`): 1.13.3
- OS (e.g. from `/etc/os-release`): Ubuntu 18.04

/assign

### LINKED ISSUE #127: Local Manager Image Build Process Broken
/kind bug

**What steps did you take and what happened:**

Follow the documentation to push custom builds of ./manager.

**What did you expect to happen:**

A successful container image can be built and pushed.

**Anything else you would like to add:**

A few issues.  This is somewhat important given quay.io/k8s is somewhat unstable, and most of the time, the images referenced in the components yaml do not exist or are broken or have broken cloud-init setup steps for kubeadm.  I'd like to build and push my own images to work around these issues and contribute fixes.

1) The override variables to point the bazel docker build and push process to a remote repository does not work properly.  The documentation says to do the following:

```bash
# Build the image.
PREFIX=quay.io/k8s \
NAME=cluster-api-azure-controller \
TAG=0.2.0-alpha.4 \
make docker-build

# Push the image.
PREFIX=quay.io/k8s \
NAME=cluster-api-azure-controller \
TAG=0.2.0-alpha.4 \
make docker-push
```

In reality, I had to:

```bash
export STABLE_DOCKER_REPO=gcr.io/cluster-api-azure-controller
export MANAGER_IMAGE_NAME=cluster-api-azure-controller
export MANAGER_IMAGE_TAG=v1.0
export DEV_DOCKER_REPO=gcr.io/cluster-api-azure-controller
make docker-build
```
2) For those unfamiliar with bazel, the [azure-plugin] documentation leads one to believe they can publish to quay.io with the latest version of Bazel.  In reality, publishing doesn't work because of lack of schema v2 support (see https://github.com/bazelbuild/rules_docker/issues/102), thus the pivot to gcr.io above.

3) The build image using `make docker-build` and `make docker-push` results in an image that lacks deployment-template.json:

```bash
$ sudo docker run --rm --entrypoint=/bin/sh -it gcr.io/cluster-api-azure-controller/cluster-api-azure-controller@sha256:a04921e530811ae5dc5021992c4640d19c8e4b53bf56de2e685048653d517c6d 
/app/cmd/manager/manager-amd64.binary.runfiles/__main__ # ls
cmd       external
/app/cmd/manager/manager-amd64.binary.runfiles/__main__ # cd /
/ # find . -name deployment-template.json
/ # find . -name manager
./app/cmd/manager
./app/cmd/manager/manager-amd64.binary.runfiles/__main__/cmd/manager
/ # exit
```
4) The `Dockerfile` (as opposed to the Bazel process) works fine. For those reading this wondering how they might work around these issues, the build process removing the heft and weight and brokeness of Bazel is as simple as the following.  I recommend either fixing `docker-build/docker-push` or simply removing them and offering instructions like the below:

```bash
make manager
docker build -t gcr.io/cluster-api-azure-controller:v1.0 .
docker push gcr.io/cluster-api-azure-controller:v1.0
```

**Environment:**

- cluster-api-provider-azure version: commit 7f482be1078f6836d8e59d4d4aa7622b4ee5ade5 / Feb 13
- Kubernetes version: (use `kubectl version`): 1.13.4
- OS (e.g. from `/etc/os-release`): Ubuntu 18.04
- Bazel:
INFO: Invocation ID: df2e00cc-bad4-4a96-9ad0-75397556659b
Build label: 0.22.0
Build target: bazel-out/k8-opt/bin/src/main/java/com/google/devtools/build/lib/bazel/BazelServer_deploy.jar
Build time: Mon Jan 28 12:58:08 2019 (1548680288)
Build timestamp: 1548680288
Build timestamp as int: 1548680288

### PR DESCRIPTION
**Pull request by justaugustus**
Signed-off-by: Stephen Augustus <saugustus@vmware.com>

**What this PR does / why we need it**:
Update developer toolbox and documentation:
- Update documentation
- Update .gitignore
- Remove skip_files in golangci config
- Bump golang version (`1.12.1`) / switch thin image to `distroless/base`
- Clean up Makefile
- Bump bazel rules_go (`0.18.1`)
- Bump kind version (`d4ff13e4808ee5a816ed92bf2d3348bda12413db`)
- Rename build/asm_shim
- Add bazel generator for kustomize
- Update example specs
- Update generate-yaml.sh
- Update cmd packages
- Add release tool
- Regenerate CRDs

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #127, fixes #108, fixes #98, fixes #107 

**Release note**:

```release-note
Developer Experience improvements
```

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -214,15 +204,15 @@ ifneq ($(FASTBUILD),y)
...
-		//pkg/cloud/azure/services/elb/mock_elbiface:go_default_library
-#cp -Rf bazel-genfiles/pkg/* pkg/
+#	//pkg/cloud/azure/services/ec2/mock_ec2iface:go_default_library \
+#	//pkg/cloud/azure/services/elb/mock_elbiface:go_default_library
+#	cp -Rf bazel-genfiles/pkg/* pkg/
 
 .PHONY: lint
 lint: dep-ensure ## Lint codebase
-	@echo If you have generated new mocks, run make copy-genmocks before linting
+	@echo If you have genereated new mocks, run make copy-genmocks before linting
```
Comment (soggiest) [inline Makefile]: SUPER NIT
Generated is spelled wrong here

File: build/controller_gen.bzl [hunk 2]
```diff
@@ -55,8 +55,8 @@ def controller_gen(name, importpath, api, visibility, deps = []):
              cd {source_package} && \\
              cp -f {project} {source_package} && \\
              GENDIR=$$(dirname {gendir})/../.. && \\
-             {controller_gen} all && \\
-             cp -fR config $$GENDIR
+            {controller_gen} all && \\
+            cp -fR config $$GENDIR
```
Comment (soggiest) [inline build/controller_gen.bzl]: It looks like the spaces were removed before these lines. Was that on purpose?

File: scripts/ci-bazel-integration.sh [hunk 3]
```diff
@@ -0,0 +1,26 @@
+#!/bin/bash
+
+# Copyright 2018 The Kubernetes Authors.
```
Comment (tariq1890) [inline scripts/ci-bazel-integration.sh]:
```suggestion
# Copyright 2019 The Kubernetes Authors.
```

File: docs/development.md [hunk 4]
```diff
@@ -48,6 +50,11 @@ cd "$(go env GOPATH)/src/sigs.k8s.io/cluster-api-provider-azure"
 make dep-ensure
 ```
 
+### Get familiar with basic concepts
+
+This provider is modeled after the upstream cluster-api project. To get familiar
+with resources, concepts and conventions refer to the [upstream gitbook](https://kubernetes-sigs.github.io/cluster-api/).
```
Comment (tariq1890) [inline docs/development.md]: SUPER DUPER NIT: can we rename it to cluster-api gitbook? Also let's move the link to the bottom of the readme.md for consistency's sake.

Comment (tariq1890) [inline docs/development.md]:
```suggestion
with the concepts and conventions, refer to the [cluster-api gitbook](https://kubernetes-sigs.github.io/cluster-api/).
```

File: cmd/clusterctl/examples/azure/credentials.yaml.template [hunk 5]
```diff
@@ -0,0 +1,12 @@
+---
+apiVersion: v1
+kind: Secret
+metadata:
+  name: manager-bootstrap-credentials
+  namespace: azure-provider-system
+type: Opaque
+data:
+  subscription-id: $AZURE_SUBSCRIPTION_ID_B64
```
Comment (tariq1890) [inline cmd/clusterctl/examples/azure/credentials.yaml.template]:
```suggestion
  subscription-id: ${AZURE_SUBSCRIPTION_ID_B64}
```

File: cmd/clusterctl/examples/azure/credentials.yaml.template [hunk 6]
```diff
@@ -0,0 +1,12 @@
...
+  tenant-id: $AZURE_TENANT_ID_B64
```
Comment (tariq1890) [inline cmd/clusterctl/examples/azure/credentials.yaml.template]:
```suggestion
  tenant-id: ${AZURE_TENANT_ID_B64}
```

File: cmd/clusterctl/examples/azure/credentials.yaml.template [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
+  client-id: $AZURE_CLIENT_ID_B64
```
Comment (tariq1890) [inline cmd/clusterctl/examples/azure/credentials.yaml.template]:
```suggestion
  client-id: ${AZURE_CLIENT_ID_B64}
```

File: cmd/clusterctl/examples/azure/credentials.yaml.template [hunk 8]
```diff
@@ -0,0 +1,12 @@
...
+  client-secret: $AZURE_CLIENT_SECRET_B64
```
Comment (tariq1890) [inline cmd/clusterctl/examples/azure/credentials.yaml.template]:
```suggestion
  client-secret: ${AZURE_CLIENT_SECRET_B64}
```

File: cmd/clusterctl/examples/azure/generate-yaml.sh [hunk 9]
```diff
@@ -83,20 +93,20 @@ while test $# -gt 0; do
         esac
 done
 
-if [ $OVERWRITE -ne 1 ] && [ -f $MACHINES_GENERATED_FILE ]; then
-  echo File $MACHINES_GENERATED_FILE already exists. Delete it manually before running this script.
+if [ $OVERWRITE -ne 1 ] && [ -f "$MACHINES_GENERATED_FILE" ]; then
```
Comment (tariq1890) [inline cmd/clusterctl/examples/azure/generate-yaml.sh]:
```suggestion
if [[ ${OVERWRITE} -ne 1 ]] && [[ -f "${MACHINES_GENERATED_FILE}" ]]; then
```

File: cmd/clusterctl/examples/azure/generate-yaml.sh [hunk 10]
```diff
@@ -83,20 +93,20 @@ while test $# -gt 0; do
...
+  echo "File $MACHINES_GENERATED_FILE already exists. Delete it manually before running this script."
```
Comment (tariq1890) [inline cmd/clusterctl/examples/azure/generate-yaml.sh]:
```suggestion
  echo "File ${MACHINES_GENERATED_FILE} already exists. Delete it manually before running this script."
```

File: cmd/clusterctl/examples/azure/generate-yaml.sh [hunk 11]
```diff
@@ -83,20 +93,20 @@ while test $# -gt 0; do
...
   exit 1
 fi
 
-if [ $OVERWRITE -ne 1 ] && [ -f $CLUSTER_GENERATED_FILE ]; then
-  echo File $CLUSTER_GENERATED_FILE already exists. Delete it manually before running this script.
+if [ $OVERWRITE -ne 1 ] && [ -f "$CLUSTER_GENERATED_FILE" ]; then
```
Comment (tariq1890) [inline cmd/clusterctl/examples/azure/generate-yaml.sh]:
```suggestion
if [[ ${OVERWRITE} -ne 1 ]] && [[ -f "${CLUSTER_GENERATED_FILE}" ]]; then
```

File: cmd/clusterctl/examples/azure/generate-yaml.sh [hunk 12]
```diff
@@ -83,20 +93,20 @@ while test $# -gt 0; do
...
+  echo "File $CLUSTER_GENERATED_FILE already exists. Delete it manually before running this script."
```
Comment (tariq1890) [inline cmd/clusterctl/examples/azure/generate-yaml.sh]:
```suggestion
  echo "File ${CLUSTER_GENERATED_FILE} already exists. Delete it manually before running this script."
```

File: cmd/clusterctl/examples/azure/generate-yaml.sh [hunk 13]
```diff
@@ -121,25 +131,37 @@ else
...
 echo "Machine SSH key generated in ${SSH_KEY_FILE}"
 
-export SSH_PUBLIC_KEY=$(cat ${SSH_KEY_FILE}.pub | base64 | tr -d '\r\n')
-export SSH_PRIVATE_KEY=$(cat ${SSH_KEY_FILE} | base64 | tr -d '\r\n')
+export SSH_PUBLIC_KEY=$(cat "${SSH_KEY_FILE}.pub" | base64 | tr -d '\r\n')
+export SSH_PRIVATE_KEY=$(cat "${SSH_KEY_FILE}" | base64 | tr -d '\r\n')
 
-$ENVSUBST < $CLUSTER_TEMPLATE_FILE > "${CLUSTER_GENERATED_FILE}"
-echo "Done generating ${CLUSTER_GENERATED_FILE}"
+if [ -z "$VNET_NAME" ]; then
```
Comment (tariq1890) [inline cmd/clusterctl/examples/azure/generate-yaml.sh]:
```suggestion
if [[ -z "${VNET_NAME}" ]]; then
```

File: docs/releasing.md [hunk 14]
```diff
@@ -1,16 +1,30 @@
 # Release Process
 
-1. Review and discuss any potential modifications to the [support matrix/policy][support-policy]
-2. Create a draft release in github
-3. Tag the repository and push the tag
+Before beginning the release process, review and discuss any potential modifications to the [support matrix/policy][support-policy].
+
+## Semi-automatic
+
+1. Make sure your repo is clean by git's standards
```
Comment (tariq1890) [inline docs/releasing.md]: Requesting a full stop//period at the end of every sentence here.

File: docs/releasing.md [hunk 15]
```diff
@@ -1,16 +1,30 @@
...
+2. If the controller has changed, it would be good to bump the version of the controller image. Look in `cmd/release/main.go`.
+3. Run `go run cmd/release/main.go -version v0.x.y`, replacing the [version][versioning], as appropriate
+4. Push the docker images that were generated with this release tool
+5. Edit the release notes and make sure the binaries uploaded return the correct version
+6. Publish release
+7. [Announce][release-announcement] the release
+
+## Manual
+
+1. Tag the repository and push the tag
```
Comment (tariq1890) [inline docs/releasing.md]: Here as well.

File: .gitignore [hunk 16]
```diff
@@ -44,3 +44,7 @@ envfile
 
 # vscode
 .vscode
+
+# Ignore output manifests
+cmd/clusterctl/examples/azure/out
```
[the remark this answers is not in the document]
Comment (justaugustus) [reply .gitignore]: I want to explicitly include this as `.../examples/` may include credentials.

File: .golangci.yml [hunk 17]
```diff
@@ -26,6 +23,8 @@ linters:
     - nakedret
     - prealloc
     - gosec
+    # TODO: Enable deadcode module once types are cleaned up
+    #- deadcode
```
[the remark this answers is not in the document]
Comment (justaugustus) [reply .golangci.yml]: We can handle this in a follow-up PR. I've added a tracking issue for this: #148

File: Dockerfile [hunk 18]
```diff
@@ -13,7 +13,7 @@
 # limitations under the License.
 
 # Build the manager binary
-FROM golang:1.11.5 as builder
+FROM golang:1.12.1 as builder
```
[the remark this answers is not in the document]
Comment (justaugustus) [reply Dockerfile]: golang version is currently pinned to the version used in Kubernetes. We're doing this to minimize any inconsistencies there.

File: docs/releasing.md (hunk 14 above)
Comment (justaugustus) [reply docs/releasing.md]: Periods are only required in bulleted lists when the bullet includes multiple sentences.

File: docs/releasing.md (hunk 15 above)
Comment (justaugustus) [reply docs/releasing.md]: Same as above.

**Comment by justaugustus** [comment]
Ready for review:
/assign @soggiest @tariq1890 @awesomenix

**Comment by awesomenix** [comment]
overall looks good, let me know once the conflicts are resolved.

**Comment by justaugustus** [comment]
Thanks @awesomenix. Conflicts fixed up!

**Pull request merged**